src/ucs/datastruct/mpool_set.c

```diff
@@ -0,0 +1,153 @@
+{
+    int i, size_log2, mpools_num;
+    int prev_idx, mps_idx, map_idx, max_idx;
+    size_t size;
+    ucs_status_t status;
+
+    if (sizes_count == 0) {
+        ucs_error("creation of empty mpool_set is not allowed");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if ((max_mp_entry_size > UCS_MPOOL_SET_MAX_SIZE) ||
+        (max_mp_entry_size == 0)) {
+        ucs_error("invalid maximal mpool size value %zu", max_mp_entry_size);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    mp_set->bitmap = 0;
+    max_idx        = UCS_MPOOL_SET_SIZE - 1;
+
+    for (i = 0; i < sizes_count; ++i) {
+        if (!ucs_is_pow2(sizes[i])) {
+            ucs_error("wrong mpool size %zu, it must be power of 2", sizes[i]);
+            return UCS_ERR_INVALID_PARAM;
+        }
+
+        if (sizes[i] > max_mp_entry_size) {
+            /* Ignore sizes bigger than maximal allowed value */
+            continue;
+        }
+
+        mp_set->bitmap |= sizes[i];
+    }
+
+    /* max_mp_entry_size is allowed to be non pow of 2. If this is the case, set
+     * the last bit in the map indicating that max_mp_entry_size is not equal to
+     * any element in sizes array.
+     */
+    if (!ucs_is_pow2(max_mp_entry_size) ||
+        !(max_mp_entry_size & mp_set->bitmap)) {
+        mp_set->bitmap |= UCS_BIT(max_idx);
+    }
+
+    mpools_num     = ucs_popcount(mp_set->bitmap);
+    mp_set->mpools = ucs_malloc(mpools_num * sizeof(*mp_set->mpools) +
+                                    priv_size,
+                                "mpools_set");
+    if (mp_set->mpools == NULL) {
+        ucs_error("failed to allocate mpools set");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    mps_idx  = 0;
+    prev_idx = max_idx;
+    ucs_for_each_bit(size_log2, mp_set->bitmap) {
+        map_idx = max_idx - size_log2;
+        size    = (map_idx == 0) ? max_mp_entry_size : UCS_BIT(size_log2);
+        status  = ucs_mpool_init(&mp_set->mpools[mps_idx], priv_size,
+                                 size + priv_elem_size, align_offset, alignment,
+                                 elems_per_chunk, max_elems, ops, name);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        /* mp_set->map is an array of pointers to memory pools. Array index is
+         * log2 of the corresponding memory pool element size. Indexation is
+         * done in the reverse order to speedup array index calculation in
+         * ucs_mpool_set_get(). Since there is no separate memory pool for
+         * every pow of 2 value, initialize all array elements in the range of
+         * [current_index, last_initalized_index).
+         * So, eventually every mp_set->map array element will point to a
+         * certain mpool with minimal element size capable of storing elements
+         * up to 2^(max_idx - index).
+         */
+        for (i = prev_idx; i >= map_idx; --i) {
+            mp_set->map[i] = &mp_set->mpools[mps_idx];
+        }
```

Comment:
maybe would be simpler to set mp_set->map[i] after ucs_for_each_bit loop instead of a nested loop

Reply:
i tried, but it looks worse. Need to determine what pointer to assign with some branches, etc

Reply:
ok